### PR TITLE
Implementation of the new targetdef element as proposed for 1.38

### PR DESCRIPTION
**Pull request by niklasharrysson**
- Added targetdef element to MaterialXCore
- Change Shader Gen to use targetsdefs for implementation identification, instead of the legacy 'language' and 'target' strings.
- Update all implementation libraries to use this new method of identification
- Update all shader generators to use this new method

### CODE REVIEW

File: libraries/targets/arnold.mtlx [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  TM & (c) 2020 Lucasfilm Entertainment Company Ltd. and Lucasfilm Ltd.
+  All rights reserved.  See LICENSE.txt for license.
+-->
+
+<materialx version="1.38">
+
+  <!-- ======================================================================== -->
+  <!-- Derived target definition for Arnold                                     -->
+  <!-- ======================================================================== -->
+  <targetdef name="arnold" inherit="genosl"/>
```
Comment (niklasharrysson) [inline libraries/targets/arnold.mtlx:12]: These are the new targetdefs elements that define a new target. A targetdef has to be loaded before any shader generation starts, in order to map implementations to nodedefs for a target application/renderer.

I put these into separate files so a client can easily select which targets to support by choosing which targetdef files to load. By default in our applications all targetdefs inside libraries/targets are loaded.

File: python/MaterialXTest/genshader.py [hunk 2]
```diff
@@ -11,7 +11,7 @@ class TestGenShader(unittest.TestCase):
     def test_ShaderInterface(self):
         doc = mx.createDocument()
 
-        libraryFolders = ["stdlib"]
+        libraryFolders = ["targets", "stdlib"]
```
Comment (niklasharrysson) [inline python/MaterialXTest/genshader.py:14]: As mentioned above the new "targets" library is loaded in, in order to have all default targetdefs loaded and supported.

File: source/MaterialXGenArnold/ArnoldShaderGenerator.cpp [hunk 3]
```diff
@@ -19,6 +20,9 @@ ArnoldShaderGenerator::ArnoldShaderGenerator()
                                       "volume_matte" };
 
     _syntax->registerReservedWords(reservedWords);
+
+    // <!-- <dielectric_brdf> -->
+    registerImplementation("IM_dielectric_brdf_" + ArnoldShaderGenerator::TARGET, ThinFilmSupport::create);
```
Comment (niklasharrysson) [inline source/MaterialXGenArnold/ArnoldShaderGenerator.cpp:25]: An unrelated fix, thin-film support was not registered properly for the Arnold generator.

File: source/MaterialXGenGlsl/Nodes/BlurNodeGlsl.h [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace MaterialX
+{
+
+/// Blur node implementation for GLSL
+class BlurNodeGlsl : public BlurNode
+{
+public:
+    static ShaderNodeImplPtr create();
+    void emitSamplingFunctionDefinition(const ShaderNode& node, GenContext& context, ShaderStage& stage) const override;
```
Comment (niklasharrysson) [inline source/MaterialXGenGlsl/Nodes/BlurNodeGlsl.h:19]: Specific blur node for GLSL to emit the right function definitions for sampling.

File: source/MaterialXGenOsl/Nodes/BlurNodeOsl.h [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace MaterialX
+{
+
+/// Blur node implementation for OSL
+class BlurNodeOsl : public BlurNode
+{
+public:
+    static ShaderNodeImplPtr create();
+    void emitSamplingFunctionDefinition(const ShaderNode& node, GenContext& context, ShaderStage& stage) const override;
```
Comment (niklasharrysson) [inline source/MaterialXGenOsl/Nodes/BlurNodeOsl.h:19]: Specific blur node for OSL to emit the right function definitions.

File: source/MaterialXGenShader/Nodes/ThinFilmNode.h [hunk 6]
```diff
@@ -29,20 +29,27 @@ class ThinFilmNode : public ShaderNodeImpl
...
     static ShaderNodeImplPtr create();
 
     void addInputs(ShaderNode& node, GenContext&) const override;
+};
 
-    /// String constants
-    static const string THINFILM_INPUT;
+/// Base class for HW source code driven BSDF nodes that support layering with thin-film.
+/// Thin-film data is added as an extra input to BSDF nodes that derive from this class.
+class HwThinFilmSupport : public HwSourceCodeNode
```
Comment (niklasharrysson) [inline source/MaterialXGenShader/Nodes/ThinFilmNode.h:47]: Split the thin-film support node into two, for OSL vs HW, since for HW we need to derive from HwSourceCodeNode.

File: source/MaterialXTest/MaterialXGenGlsl/GenGlsl.cpp [hunk 7]
```diff
@@ -103,7 +103,7 @@ TEST_CASE("GenShader: Bind Light Shaders", "[genglsl]")
 
     mx::FileSearchPath searchPath;
     searchPath.append(mx::FilePath::getCurrentPath() / mx::FilePath("libraries"));
-    loadLibraries({ "stdlib", "pbrlib", "lights" }, searchPath, doc);
+    loadLibraries({ "targets", "stdlib", "pbrlib", "lights" }, searchPath, doc);
```
Comment (niklasharrysson) [inline source/MaterialXTest/MaterialXGenGlsl/GenGlsl.cpp:106]: Again, loading in the default targetdefs along with the standard libraries.

File: source/MaterialXTest/MaterialXGenShader/GenShader.cpp [hunk 8]
```diff
@@ -98,116 +98,6 @@ TEST_CASE("GenShader: TypeDesc Check", "[genshader]")
     REQUIRE_THROWS(mx::TypeDesc::get("bar"));
 }
 
-TEST_CASE("GenShader: OSL Reference Implementation Check", "[genshader]")
```
Comment (niklasharrysson) [inline source/MaterialXTest/MaterialXGenShader/GenShader.cpp:101]: Remove this test as the legacy OSL implementations are outdated and are going to be removed anyway, and so I didn't want to spend time on updating this.

File: libraries/targets/arnold.mtlx (hunk 1 above)
Comment (bernardkwok) [reply libraries/targets/arnold.mtlx:12]: Just curious if we should add glsl, osl, mdl, genosl, targets into stdlib.mtlx so it always get's loaded without any change to existing integrations? 
Then only the non-core ones would be added to the targets directory?

File: source/MaterialXCore/Definition.cpp [hunk 9]
```diff
@@ -10,6 +10,38 @@
...
+            if (!interface->isA<Implementation>() ||
+                !targetStringsMatch(interface->getTarget(), target) ||
+                !nodedef.isVersionCompatible(interface))
+            {
+                continue;
+            }
+            return interface;
+        }
+
+        // Search for a node graph match if no implementation match was found.
```
[the remark this answers is not in the document]
Comment (niklasharrysson) [reply source/MaterialXCore/Definition.cpp:29]: This split into two loops was done in the old code, so I didn't change that. I think it's because we consider source code or native implementations to have higher priority than node graph implementations. And I think that makes sense, a native or source code implementation is more specific so should have higher priority.

File: libraries/targets/arnold.mtlx (hunk 1 above)
Comment (niklasharrysson) [reply libraries/targets/arnold.mtlx:12]: It's a bit unfortunate that an existing integration now has to make a change to make sure the appropriate target support gets loaded. However I think it might be best if that happens, to force a client to select which targets to support. 

Because then we can make sure to only load in implementations that are really needed. For example, for a GLSL client it's a waste of performance to have all implementations for glsl, osl and mdl loaded when only glsl is used. Not only in the first pass when it's loaded in, but having all these loaded in the map for each successive query of implementations.

I think a "breaking" change like this is OK since it's a move to a new version 1.38, and we don't have a wide spread use of MaterialX anyway yet. So better to make this change now than later.

Keeping it like this also make target loading consistent, and avoid classifying some targets as "core" and others as not as important. It's better if the API client can make that decision.

Comment (niklasharrysson) [reply libraries/targets/arnold.mtlx:12]: The fact that I load in all targetdefs that are available in the "targets" folder is just a simplification for now. We can later make this more specific for our various back-ends and only load in the targets that are needed.

File: source/MaterialXCore/Definition.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (bernardkwok) [reply source/MaterialXCore/Definition.cpp:29]: Ok. I misread the diff then. Thanks.

**Review by bernardkwok — APPROVED** [review]

**Review by jstone-lucasfilm — APPROVED** [review]
Thanks @niklasharrysson, and this looks good to me!
